scripts/bert/finetune_classifier.py

```diff
@@ -111,60 +142,89 @@ def evaluate():
     for _, v in model.collect_params('.*beta|.*gamma|.*bias').items():
         v.wd_mult = 0.0
-
+    # Collect differentiable parameters
     for p in model.collect_params().values():
         if p.grad_req != 'null':
-            differentiable_params.append(p)
+            params.append(p)
+    # Set grad_req if gradient accumulation is required
+    if accumulate:
+        for p in model.collect_params().values():
+            p.grad_req = 'add'
 
     for epoch_id in range(args.epochs):
         metric.reset()
         step_loss = 0
-
+        tic = time.time()
         for batch_id, seqs in enumerate(bert_dataloader):
-            step_num += 1
+            # set grad to zero for gradient accumulation
+            if accumulate:
+                if batch_id % accumulate == 0:
+                    model.collect_params().zero_grad()
+                    step_num += 1
+            else:
+                step_num += 1
+            # learning rate schedule
             if step_num < num_warmup_steps:
                 new_lr = lr * step_num / num_warmup_steps
             else:
                 offset = (step_num - num_warmup_steps) * lr / (num_train_steps - num_warmup_steps)
                 new_lr = lr - offset
             trainer.set_learning_rate(new_lr)
+            # forward and backward
             with mx.autograd.record():
                 input_ids, valid_length, type_ids, label = seqs
                 out = model(input_ids.as_in_context(ctx), type_ids.as_in_context(ctx),
                             valid_length.astype('float32').as_in_context(ctx))
                 ls = loss_function(out, label.as_in_context(ctx)).mean()
             ls.backward()
-            grads = [p.grad(ctx) for p in differentiable_params]
-            gluon.utils.clip_global_norm(grads, 1)
-            trainer.step(1)
+            # update
+            if not accumulate or (batch_id + 1) % accumulate == 0:
+                grads = [p.grad(ctx) for p in params]
+                gluon.utils.clip_global_norm(grads, 1)
+                trainer.step(accumulate if accumulate else 1)
             step_loss += ls.asscalar()
             metric.update([label], [out])
-            if (batch_id + 1) % (args.log_interval) == 0:
+            if (batch_id + 1) % log_interval == 0:
                 logging.info('[Epoch {} Batch {}/{}] loss={:.4f}, lr={:.7f}, acc={:.3f}'
                              .format(epoch_id, batch_id + 1, len(bert_dataloader),
-                                     step_loss / args.log_interval,
+                                     step_loss / log_interval,
                                      trainer.learning_rate, metric.get()[1]))
                 step_loss = 0
         mx.nd.waitall()
         evaluate()
+        toc = time.time()
+        logging.info('Time cost={:.1f}s'.format(toc - tic))
+        tic = toc
 
 if __name__ == '__main__':
+    pool_type = os.environ.get('MXNET_GPU_MEM_POOL_TYPE', '')
```

Comment:
Given that static_alloc is on in the model, mem pool type would not be affecting the input and output arrays. In this case, you may want to tune the `MXNET_GPU_MEM_POOL_ROUND_LINEAR_CUTOFF=x` so that 2^x is on the same scale as the mode of the input output array sizes.

Reply:
It might be too much for users to understand how to set it properly... Would it be better to remove this option for now to keep the example simple?